src/ucs/datastruct/ptr_array.h

```diff
@@ -125,8 +124,100 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
  * Iterate over all valid elements in the array.
  */
 #define ucs_ptr_array_for_each(_var, _index, _ptr_array) \
-    for (_index = 0; _index < (_ptr_array)->size; ++_index) \
-         if (!__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index]))) \
+    unsigned elements = (_ptr_array)->n_occupied;\
+    unsigned is_element = 0;\
+    for (_index = 0; (elements > 0) && (_index < (_ptr_array)->size); elements -= is_element, ++_index)\
+       if ((is_element = ucs_likely(!__ucs_ptr_array_is_free(_var = (void *)((_ptr_array)->start[_index])))))\
+
+
+/*
+   Locked interface
+*/
+
+/* Locked ptr array */
+typedef struct ucs_ptr_array_locked {
+    ucs_ptr_array_t super;
+    ucs_recursive_spinlock_t lock;
+} ucs_ptr_array_locked_t;
+
+/**
+ * Locked array init
+ * @param  index    index of slot
+ * @param  new_val  value to put into slot given by index
+ * @return old value of the slot
+ */
+void ucs_ptr_array_locked_init(ucs_ptr_array_locked_t *locked_ptr_array, uint32_t init_placeholder,
+                        const char *name);
+
+
+/**
+ * Cleanup the locked array.
+ * All values should already be removed from it.
+ */
+void ucs_ptr_array_locked_cleanup(ucs_ptr_array_locked_t *locked_ptr_array);
+
+
+/**
+ * Insert a pointer to the locked array.
+ *
+ * @param value        Pointer to insert. Must be 8-byte aligned.
+ * @param placeholder  Filled with placeholder value.
+ * @return             The index to which the value was inserted.
+ *
+ * Complexity: amortized O(1)
+ *
+ * Note: The array will grow if needed.
+ */
+unsigned ucs_ptr_array_locked_insert(ucs_ptr_array_locked_t *locked_ptr_array, void *value,         
+                   uint32_t *placeholder_p);
 
 
+/**
+ * Remove a pointer from the locked array.
+ *
+ * @param index        Index to remove from.
+ * @param placeholder  Value to put in the free slot.
+ *
+ * Complexity: O(1)
+ */
+void ucs_ptr_array_locked_remove(ucs_ptr_array_locked_t *locked_ptr_array, unsigned index,
+                          uint32_t placeholder);
+
+
+/**
+ * Replace pointer in the locked array
+ * @param  index    index of slot
+ * @param  new_val  value to put into slot given by index
+ * @return old value of the slot
+ */
+void *ucs_ptr_array_locked_replace(ucs_ptr_array_locked_t *locked_ptr_array,
+	unsigned index, void *new_val);
+
+/* Acquire ptr_array lock */
```

Comment:
I assume we need those for the for-each? if so, we should document this (otherwise unclear when it should be used).

Reply:
Added comments to both locked_lookup() and locked_for_each() functions prototype:


/**
 * Iterate over all valid elements in the locked array.
 * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
 * and ucs_ptr_array_locked_release_lock() functions.
 */
#define ucs_ptr_array_locked_for_each(_var, _index, _locked_ptr_array) \
    ucs_ptr_array_for_each(_var, _index, &(_locked_ptr_array)->super)